### PR TITLE
Feature/update wp user profile

### LINKED ISSUE #56: Webhook to Search new WooComerce user in CRM to get automatic discount
I open ticket here to publicly document the process.

Used the following plugins in Wordpress/Woocomerce in order to apply an automatic discount for the clients that appear in the CRM as partners:

https://www.webtoffee.com/how-apply-woocommerce-discount-based-user-roles/
https://wordpress.org/plugins/wt-smart-coupons-for-woocommerce/

https://wisdmlabs.com/blog/apply-discount-coupon-automatically-woocommerce/
https://wordpress.org/plugins/woocommerce-auto-added-coupons/#screenshots

Documentation used:
http://wiki.sinergiacrm.org/images/Sugar_developer_guide_6.5.zip
http://support.sugarcrm.com/04_Knowledge_Base/
https://community.sugarcrm.com/sugarcrm

### PR DESCRIPTION
**Pull request by numeropi**
- Adds a task to update the Wordpress user's profile to have the new role defined in the env var "WP_PARTNER_ROLE" to mark them as customer-partner.

closes #56

### CODE REVIEW

File: wordpress/user.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import wordpress
+from wordpress.client import APIClient
+
+
+class WPUser:
+    email = None
+    roles = None
+    user_id = None
```
Comment (sauloperez) [inline wordpress/user.py:8]: class attributes won't allow us to have multiple instances of user. They should really be instance attributes.

File: wordpress/user.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        self.roles = set()
+        self.client = client
+
+
+    def fetch(self, email):
+        response = self.client.get_request(f'users/?search={email}')
+        body = response.json()
+        if len(body):
+            user_id = body[0]['id']
+            response = self.client.post_request(f'users/{user_id}')
```
Comment (sauloperez) [inline wordpress/user.py:21]: really? WP makes us issue a POST request to get a user? :see_no_evil:

File: wordpress/user.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+            body = response.json()
+            email_profile = body['email']
+            if email == email_profile:
```
Comment (sauloperez) [inline wordpress/user.py:24]: isn't this check redundant? If at line 17 we fetched by the email what we know for sure is that emails match.

File: wordpress/user.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    
+    def __init__(self, client=APIClient(), logger=wordpress.logger):
+        self.logger = logger
+        self.roles = set()
+        self.client = client
+
+
+    def fetch(self, email):
+        response = self.client.get_request(f'users/?search={email}')
+        body = response.json()
```
Comment (sauloperez) [inline wordpress/user.py:18]: Parsing the response is a transport-level detail that belongs to `get_request` IMO. Moving it there would also remove duplication.

File: wordpress/client.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
+import requests
+import wordpress
+
+class APIClient:
+    api_path = "wp-json/wp/v2"
+    user = wordpress.wp_user
+    password = wordpress.wp_password
+    api_url = f'{wordpress.wp_host}/{api_path}'
```
Comment (sauloperez) [inline wordpress/client.py:8]: again, I think restricting this class to only have a single instance it's an unnecessary constraint it may cause us trouble in the future.

File: wordpress/user.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    roles = None
+    user_id = None
+    
+    def __init__(self, client=APIClient(), logger=wordpress.logger):
+        self.logger = logger
+        self.roles = set()
+        self.client = client
+
+
+    def fetch(self, email):
```
Comment (sauloperez) [inline wordpress/user.py:16]: I see that not having the user id persisted in the DB is making this rather inefficient. And I see this will scale quickly as we implement more use cases. We need to start thinking about persistence.

File: wordpress/user.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        self.roles = body['roles']
+        self.username = body['username']
+        self.email = body['email']
+        self.user_id = body['id']
+
+    def update_as_partner(self):
+        if wordpress.wp_partner_role not in self.roles:
+            new_roles = ",".join(self.roles + wordpress.wp_partner_role)
+            response = self.client.post_request(f'users/{self.user_id}', data={'roles': new_roles})
+            response.raise_for_status()
```
Comment (sauloperez) [inline wordpress/user.py:40]: this also belongs to the client and consumers shouldn't be responsible for it.

File: wordpress/user.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        return self
+
+    def _convert_to_resource(self, response):
+        body = response.json()
+        self.roles = body['roles']
+        self.username = body['username']
+        self.email = body['email']
+        self.user_id = body['id']
+
+    def update_as_partner(self):
```
Comment (sauloperez) [inline wordpress/user.py:36]: I would rather keep this user-resource class implement CRUD only. Implement a set of fetch, update, create, delete methods is what we need here. Then, this more use-case-specific logic can go into the Celery task itself. Only if we see the need to support more advanced methods beyond simple CRUD, I would consider adding them here in the future.

TL;DR: I turn this into (disclaimer, this is pseudo-codish):

```py
def update(self, **kwargs):
  response = self.client.post_request(f'users/{self.user_id}', data=locals())
  self._convert_to_resource(response)
```

and then on `update_user_as_partner`, which is at one level of abstraction higher:

```py
def update_user_as_partner(email):
  wp_user = WPUser().fetch(email)
  if wordpress.wp_partner_role not in self.roles:
    new_roles = ",".join(self.roles + wordpress.wp_partner_role)
    wp_user.update(roles=new_roles)
```
:ok_hand:

**Comment by sauloperez** [comment]
have you investigated potential Python API clients we could use? I searched both in Google and Github and didn't find any clear candidates.

File: wordpress/user.py (hunk 2 above)
Comment (numeropi) [reply wordpress/user.py:21]: No, sorry, same effect with a GET... well catched, i copied the post here since it's needed to update the roles... and I saw that an empty post also was getting the user.

File: wordpress/user.py (hunk 3 above)
Comment (numeropi) [reply wordpress/user.py:24]: Totally, not sure what i was thinking.

File: wordpress/user.py (hunk 8 above)
Comment (numeropi) [reply wordpress/user.py:36]: Looks good!

File: wordpress/user.py (hunk 7 above)
Comment (numeropi) [reply wordpress/user.py:40]: Actually i already had them inside the requests, not sure why this one ended up here too.

File: wordpress/client.py (hunk 5 above)
Comment (numeropi) [reply wordpress/client.py:8]: Definetely should be instance attributes. thanks

File: wordpress/user.py (hunk 6 above)
Comment (numeropi) [reply wordpress/user.py:16]: You mean saving the WpUsers with id+email in the db for future usercases? it makes sense. For the moment since this is only trigerred when new wordpress user Ids are created (new Woocomerce customers), this use case won't reuse any previous user_id.

**Comment by numeropi** [comment]
> have you investigated potential Python API clients we could use? I searched both in Google and Github and didn't find any clear candidates but I would really like to avoid reinventing the wheel

Yeah, i made some research, this one uses the Oauth plugin that is more secure for authentication, https://github.com/derwentx/wp-api-python

and https://github.com/stylight/python-wordpress-json, but this one I saw it too focused on WP classes(posts, comments, etc)... not really focused in WooComerce.

File: wordpress/user.py (hunk 2 above)
Comment (numeropi) [reply wordpress/user.py:21]: Now testing I remember why I discarded GET... it does not return the 'roles' attribute of the user. I will investigate a bit more if this can be easily get with a get and some parameters to avoid using empty POST.

File: wordpress/client.py (hunk 5 above)
Comment (numeropi) [reply wordpress/client.py:8]: Should be also changed probably in Koiki wcfmmp.py

File: wordpress/user.py (hunk 2 above)
Comment (numeropi) [reply wordpress/user.py:21]: Discovered that even if the WP api allows multiple roles for a user: 
image

The Wordpress admin panel user editor does not, and this may create management problems since it won't show if the user has a partner role and if saved again using the wordpress admin it may remove it.
image

Maybe we should just make sure that the current role is "customer" (to not replace an "admin" role for example), and replace "customer" which is the role used by woocomerce newly created users.

File: wordpress/user.py (hunk 6 above)
Comment (sauloperez) [reply wordpress/user.py:16]: yes, we can still afford it but it is very clunky to search the user by email to then fetch it again when its id probably comes within the new customer webhook's body.

**Comment by sauloperez** [comment]
> and https://github.com/stylight/python-wordpress-json, but this one I saw it too focused on WP classes(posts, comments, etc)... not really focused in WooComerce.

I'm afraid we'll need to use two separate API clients even though Woocommerce's API is built on top of WP's. They have a different set of REST resources :shrug:

File: .gitignore [hunk 9]
```diff
@@ -5,3 +5,4 @@ htmlcov/
 tags
 session.json
 data.json
+label.pdf
```
Comment (sauloperez) [inline .gitignore:8]: oops, I forgot :joy:

File: wordpress/tests/test_wpuser.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+            responses.GET,
+            f"https://test_wp_host/wp-json/wp/v2/users/{self.user_id}?context=edit",
+            status=200,
+            json={
+                "id": self.user_id,
+                "username": "testusername",
+                "email": self.email,
+                "roles": ["testrole"],
+            },
+        )
```
Comment (sauloperez) [inline wordpress/tests/test_wpuser.py:38]: I don't mean to change this now but we have a bit of a mess with response stubs. See https://github.com/coopdevs/lazona_connector/issues/13. But I was also trying to get the latest changes on https://github.com/coopdevs/lazona_connector/pull/64. We have to stick to one.

**Review by sauloperez — APPROVED** [review]

File: wordpress/user.py (hunk 6 above)
Comment (numeropi) [reply wordpress/user.py:16]: I agree.

**Comment by numeropi** [comment]
> > and https://github.com/stylight/python-wordpress-json, but this one I saw it too focused on WP classes(posts, comments, etc)... not really focused in WooComerce.
> 
> I'm afraid we'll need to use two separate API clients even though Woocommerce's API is built on top of WP's. They have a different set of REST resources

I think in long term if we want to move to Oauth we could use https://github.com/derwentx/wp-api-python is not python-wrapping the resources REST resources but let you send any endpoint you want.
